crud/src/main/java/com/redhat/lightblue/mediator/DefaultMetadataResolver.java

```diff
@@ -183,43 +185,60 @@ private void addFieldRoles(Set<String> roles, FieldTreeNode node) {
-        // store root entity metadata
-        metadataMap.put(entityName, emd);
+    private void initMetadataMap(EntityMetadata emd, String rootEntityName, String rootRequestVersion) {
+        // We keep two lists:
+        //   toProcess: When metadata is loaded, it is put into this list to be processed later
+        //   deferred: When metadata cannot be loaded because its version is not known, ifs name is added here
+        ArrayDeque<EntityMetadata> toProcess=new ArrayDeque<>();
         Set<String> deferred = new HashSet<>();
+
+        toProcess.addLast(emd);
+        while(!toProcess.isEmpty()||!deferred.isEmpty()) {
+            EntityMetadata metadata=toProcess.pollFirst();
+            if(metadata!=null) {
+                processMd(metadata,toProcess,deferred);
+                deferred.remove(metadata.getName());
+                metadataMap.put(metadata.getName(),metadata);
+            }
+            while(!deferred.isEmpty()) {
+                String e=deferred.iterator().next();
+                deferred.remove(e);
+                if(!metadataMap.containsKey(e)) {
+                    EntityMetadata defEmd = md.getEntityMetadata(e, null);
+                    if (defEmd == null) {
+                        throw Error.get(CrudConstants.ERR_UNKNOWN_ENTITY, e + ":default");
+                    }
+                    toProcess.addLast(defEmd);
+                }
+            }
+        }
+    }
+
+    private void processMd(EntityMetadata emd,Deque<EntityMetadata> toProcess,Set<String> deferred) {
         FieldCursor cursor = emd.getFieldCursor();
         while (cursor.next()) {
             FieldTreeNode node = cursor.getCurrentNode();
             if (node instanceof ReferenceField)
-                validateRefField(emd, entityName, rootRequestVersion, deferred, (ReferenceField) node);
+                validateRefField(emd, toProcess, deferred, (ReferenceField) node);
         }
-
-        // If there are entitis for which didn't find specific versions, use default
-        deferred.stream().filter(n->!metadataMap.containsKey(n)).forEach(n -> {
-                EntityMetadata defEmd = md.getEntityMetadata(n, null);
-                if (defEmd == null) {
-                    throw Error.get(CrudConstants.ERR_UNKNOWN_ENTITY, n + ":default");
-                }
-                metadataMap.put(n, defEmd);
-            });
     }
     
-    private void validateRefField(EntityMetadata emd, String entityName, String rootRequestVersion, Set<String> deferred, ReferenceField ref) {
+    private void validateRefField(EntityMetadata emd, Deque<EntityMetadata> toProcess, Set<String> deferred, ReferenceField ref) {
         String name = ref.getEntityName();
         String ver = ref.getVersionValue();
-        EntityMetadata refMd = null;
+        EntityMetadata refMd = metadataMap.get(name);
         if (ver == null) { // if null, default version
-            deferred.add(name);
+            if(refMd==null)
```

Comment:
We defer loading if ver==null and it is not already loaded. There is an else to (if ver==null) part where the actual loading happens, that's why this can't be consolidated into a single condition.